consensus/scheme/rolldpos/rolldposctx_test.go

```diff
@@ -265,6 +260,85 @@ func TestCheckBlockProposer(t *testing.T) {
 	require.NoError(rctx.CheckBlockProposer(51, bp, en))
 }
 
+func TestNotProducingMultipleBlocks(t *testing.T) {
+	require := require.New(t)
+	cfg := config.Default
+	b, sf, _, rp, pp := makeChain(t)
+	c := clock.New()
+	cfg.Genesis.BlockInterval = time.Second * 20
+	rctx, err := newRollDPoSCtx(
+		consensusfsm.NewConsensusConfig(cfg),
+		config.Default.DB,
+		true,
+		time.Second,
+		true,
+		b,
+		rp,
+		nil,
+		func(epochnum uint64) ([]string, error) {
+			re := protocol.NewRegistry()
+			if err := rp.Register(re); err != nil {
+				return nil, err
+			}
+			tipHeight := b.TipHeight()
+			ctx := protocol.WithBlockchainCtx(
+				protocol.WithRegistry(context.Background(), re),
+				protocol.BlockchainCtx{
+					Genesis: config.Default.Genesis,
+					Tip: protocol.TipInfo{
+						Height: tipHeight,
+					},
+				},
+			)
+			tipEpochNum := rp.GetEpochNum(tipHeight)
+			var candidatesList state.CandidateList
+			var addrs []string
+			var err error
+			switch epochnum {
+			case tipEpochNum:
+				candidatesList, err = pp.Delegates(ctx, sf)
+			case tipEpochNum + 1:
+				candidatesList, err = pp.NextDelegates(ctx, sf)
+			default:
+				err = errors.Errorf("invalid epoch number %d compared to tip epoch number %d", epochnum, tipEpochNum)
+			}
+			if err != nil {
+				return nil, err
+			}
+			for _, cand := range candidatesList {
+				addrs = append(addrs, cand.Address)
+			}
+			return addrs, nil
+		},
+		"",
+		identityset.PrivateKey(10),
+		c,
+		config.Default.Genesis.BeringBlockHeight,
+	)
+	require.NoError(err)
+	require.NotNil(rctx)
+	require.NoError(rctx.Start(context.Background()))
+	defer rctx.Stop(context.Background())
+
+	res, err := rctx.Proposal()
+	require.NoError(err)
+	ecm, ok := res.(*EndorsedConsensusMessage)
+	require.True(ok)
+	hash1, err := ecm.Document().Hash()
+	require.NoError(err)
+	height1 := ecm.Height()
+
+	res2, err := rctx.Proposal()
```

Comment:
is it possible that the proposer has been switched between these two ".Proposal()" calls?

Reply:
In this unit test, I assumed the case that the same proposer's node crashes and restart, so call Proposal() again. And make sure using cache from eManager.

Reply:
So didn't consider proposer switching or round update

Reply:
If the proposer has been switched in the middle, the changed proposer will not be the proposer of this round, so will skip it.